### ISSUE TITLE
RTSP camera streaming leaks NVENC encode sessions on writer detach → NVENCSTATUS=21, streams dead until full process restart

### ISSUE THREAD

**Issue by martinfg** [isaac-sim/IsaacSim#743]
### Description

# RTSP camera streaming leaks NVENC encode sessions on writer detach → NVENCSTATUS=21, streams dead until full process restart

## Summary

`isaacsim.streaming.rtsp` (H.264 mode) does **not release the NVENC encode session when the writer detaches**. Every attach opens new encode session(s); detach tears down the RTSP server and detaches the annotator, but the underlying NVENC session (`omni.replicator.nv` → `NvEncSession` / `OgnPostRenderVarHwEncode`) is not freed. On a consumer GeForce card the sessions accumulate across attach/detach (i.e. timeline stop→play, or programmatic re-attach) until the concurrent-session cap is hit, after which `nvEncOpenEncodeSessionEx` fails with `NVENCSTATUS=21` (`NV_ENC_ERR_INCOMPATIBLE_CLIENT_KEY`). The writers then enter the "failed" state and the streams stay dead **until the whole Isaac Sim process is restarted** — a timeline restart does not recover them.

This makes RTSP camera streaming unusable for any iterative workflow that stops/replays the timeline.

## Environment

- Isaac Sim **6.0.0** (container `nvcr.io/nvidia/isaac-sim:6.0.0`), app `isaacsim.exp.full.streaming.kit`, headless (`--no-window`)
- Extensions: `isaacsim.streaming.rtsp`, `omni.replicator.nv` **1.1.1**
- GPU: **NVIDIA GeForce RTX 5090**, driver **580.105.08**, CUDA 13.0
- Encoding: default `h264` (NVENC)

## Minimal reproduction (no action graph, no `RTSPCameraHelper`)

Run in the Script Editor. Fresh empty stage, 3 plain cameras, script the writers directly, attach/detach 6 times:

```python
import asyncio, carb, omni.kit.app, omni.usd, omni.timeline
import omni.replicator.core as rep
from pxr import UsdGeom
from isaacsim.streaming.rtsp import RTSPStreamWriter
from isaacsim.streaming.rtsp.impl.render_var_utils import ensure_render_var_on_product

async def repro():
    ctx = omni.usd.get_context()
    await ctx.new_stage_async()
    stage = ctx.get_stage()
    cams = ["/World/Cam0", "/World/Cam1", "/World/Cam2"]
    for c in cams:
        UsdGeom.Camera.Define(stage, c)
    omni.timeline.get_timeline_interface().play()
    app = omni.kit.app.get_app()
    rps = [rep.create.render_product(c, (1280, 720)) for c in cams]
    for rp in rps:
        ensure_render_var_on_product(stage, rp.path, "LdrColor", "h264")
    for i in range(6):
        writers = [RTSPStreamWriter(port=8600 + j, mountPath="/cam", encoding="h264",
                                    width=1280, height=720) for j in range(3)]
        for w, rp in zip(writers, rps):
            w.attach([rp])
        for _ in range(90):
            await app.next_update_async()
        up = sum(w._server is not None for w in writers)   # server starts on first frame
        carb.log_warn(f"[repro] cycle {i}: servers_up={up}/3")
        for w in writers:
            w.detach()
        for _ in range(40):
            await app.next_update_async()

asyncio.ensure_future(repro())
```

## Expected

`detach()` releases the NVENC session (the docs state the server is torn down and the port released on detach, and that you may "detach and re-attach as needed"). Every cycle should print `servers_up=3/3` and the run should be repeatable indefinitely.

## Actual

Sessions are never freed; they accumulate ~3 per cycle against the GeForce concurrent-session cap (~8 here):

```
[repro] cycle 0: servers_up=3/3
[repro] cycle 1: servers_up=3/3
[repro] cycle 2: servers_up=2/3
[repro] cycle 3: servers_up=0/3
[repro] cycle 4: servers_up=0/3
[repro] cycle 5: servers_up=0/3
```

with the log flooded by:

```
[Error] [omni.replicator.nv.plugin] NvEncSession: nvEncOpenEncodeSessionEx failed (NVENCSTATUS=21)
[Error] [omni.replicator.nv.plugin] OgnPostRenderVarHwEncode: failed to initialize NVENC session.
```

Once exhausted, the writers stay failed; only restarting the Isaac Sim process recovers streaming.

## Also reproduces via the declarative path

With the `RTSPCameraHelper` action-graph nodes baked into a scene (3 cameras), pressing **Play/Stop** repeatedly breaks on the **3rd Play** with the same errors — confirming it's independent of whether the writer is driven by the OmniGraph node or scripted directly.

## Notes

- `nvidia-smi --query-gpu=encoder.stats.sessionCount` (and `-q -d ENCODER_STATS`) reports **0 active sessions throughout**, even mid-failure — the counter does not attribute these sessions, so it can't be used to observe the leak; the reliable signal is the `NVENCSTATUS=21` onset.
- Secondary issue: a render product configured at **320×180** is rejected by NVENC as an invalid extent and silently bumped to **320×256** (`omni.kit.livestream.rtsp.plugin: Processing static resize ... 320x180 ... invalid ... adjusted to 320x256`). The encoded resolution then differs from what was requested.

## Impact & workaround

- **Impact:** RTSP camera streaming cannot survive any timeline stop/play or programmatic re-attach; recovery requires a full process restart.
- **Workaround:** keep the timeline **playing** and never detach/re-attach — with no re-attach the leak never triggers (verified stable over continuous streaming).

### Isaac Sim version

6.0.0 (Docker)

### Operating System (OS)

Ubuntu 24.04.3 LTS

### GPU Name

RTX 5090

### GPU Driver and CUDA versions

Driver 580.105.08, CUDA 13.0

### Logs

[Error] [omni.replicator.nv.plugin] NvEncSession: nvEncOpenEncodeSessionEx failed (NVENCSTATUS=21)
[Error] [omni.replicator.nv.plugin] OgnPostRenderVarHwEncode: failed to initialize NVENC session.

**Comment by zwdoescode** [comment]
@martinfg our latest Isaac Sim is 6.0.1. Could you please try and see if you can replicate this issue? Thanks!

**Comment by martinfg** [comment]
Tested on 6.0.1 (isaacsim.streaming.rtsp 0.1.4). Still leaks, identically to 6.0.0.

Same machine, same GPU, same repro back to back — 3 cameras @ 1280x720 h264, attach / 90 frames / detach per cycle. Both versions produce the identical curve: cycles 0 and 1 bring up all 3 streams with zero failures, cycle 2 drops to 2/3 with 90 failures, and from cycle 3 onward it's 0/3 with 270 failures every cycle. Six encode sessions (2 cycles × 3 cameras) are consumed and never returned; from cycle 3 the process can't encode at all and never recovers — only a full restart clears it. So writer.detach() doesn't release the NVENC session.

```
[Error] [omni.replicator.nv.plugin] NvEncSession: nvEncOpenEncodeSessionEx failed (NVENCSTATUS=21)
[Error] [omni.replicator.nv.plugin] OgnPostRenderVarHwEncode: failed to initialize NVENC session.
```

Failure counts corroborate the stream counts exactly: 90 = one camera failing on each of 90 frames, 270 = all three.

**Comment by zwdoescode** [comment]
Thanks for retesting on 6.0.1 and for the failure counts. 

I ran the same 3-camera, 6-cycle H.264 attach/detach repro on a local 6.0.1-rc.7 build with an RTX 5000 Ada. That system did not hit the failure: all six cycles stayed at `3/3` streams and no `NVENCSTATUS=21` appeared. The local GPU is not equivalent to your RTX 5090 GeForce setup, so this does not invalidate your report.

Source and runtime graph inspection point to the H.264 pre-encode path: `RTSPStreamWriter(encoding="h264")` uses Replicator `LdrColor` with `compression="h264"`, which activates `omni.replicator.nv.OgnPostRenderVarHwEncode`. The RTSP writer and graph node do call detach/cleanup, so the likely failing owner is the lower-level NVENC session teardown in that HW encode path.

Current workaround remains to keep the stream attached and the timeline playing. One additional workaround worth trying is raw mode, which avoids `OgnPostRenderVarHwEncode`: use `RTSPStreamWriter(..., encoding="raw")` or set `RTSPCameraHelper.inputs:useRawEncoding=True`. Please note raw mode has different performance/metadata behavior.

I also reproduced the secondary 320x180 behavior locally: the RTSP backend adjusts it to 320x256. That should be tracked separately from the NVENC session leak.

I will create internal ticket for your issue. Thanks

Internal ticket: 6637539
